infrastructure/comms/src/connection/net_address/net_addresses.rs

```diff
@@ -0,0 +1,253 @@
+pub struct NetAddresses {
+    addresses: Vec<NetAddressWithStats>,
+}
+
+impl NetAddresses {
+    /// Constructs a new list of addresses with usage stats from a list of net addresses
+    pub fn new(addresses: Vec<NetAddressWithStats>) -> NetAddresses {
+        NetAddresses { addresses }
+    }
+
+    /// Constructs a new list of addresses with usage stats from a single net address
+    pub fn from_net_address(address: NetAddress) -> NetAddresses {
+        NetAddresses {
+            addresses: vec![NetAddressWithStats::from_net_address(address)],
+        }
+    }
+
+    /// Provides the date and time of the last successful communication with this peer
+    pub fn last_seen(&self) -> Option<DateTime<Utc>> {
+        let mut latest_valid_datetime: Option<DateTime<Utc>> = None;
+        for curr_address in &self.addresses {
+            if curr_address.last_seen.is_some() {
+                if latest_valid_datetime.is_none() {
+                    latest_valid_datetime = curr_address.last_seen;
+                } else if latest_valid_datetime.unwrap() < curr_address.last_seen.unwrap() {
+                    latest_valid_datetime = curr_address.last_seen;
+                }
+            }
+        }
+        latest_valid_datetime
+    }
+
+    /// Adds a new net address to the peer if it doesn't yet exist
+    pub fn add_net_address(&mut self, net_address: NetAddress) -> Result<(), NetAddressError> {
+        let mut found_flag = false;
+        for curr_address in &self.addresses {
+            if curr_address.net_address == net_address {
+                found_flag = true;
+                break;
+            }
+        }
+        if !found_flag {
+            self.addresses.push(NetAddressWithStats::from_net_address(net_address));
+            Ok(())
+        } else {
+            Err(NetAddressError::DuplicateAddress)
+        }
+    }
+
+    /// Finds and returns the highest priority net address until all connection attempts for each net address have been
+    /// reached
+    pub fn get_net_address(&mut self) -> Result<NetAddress, NetAddressError> {
+        if self.addresses.len() >= 1 {
+            let mut any_reachable = false;
+            for curr_address in &mut self.addresses {
+                if curr_address.connection_attempts < MAX_CONNECTION_ATTEMPTS {
+                    any_reachable = true;
+                    break;
+                }
+            }
+            if any_reachable {
+                self.addresses.sort();
+                Ok(self.addresses[0].net_address.clone())
+            } else {
+                Err(NetAddressError::ConnectionAttemptsExceeded)
+            }
+        } else {
+            Err(NetAddressError::NoValidAddresses)
+        }
+    }
+
+    /// The average connection latency of the provided net address will be updated to include the current measured
+    /// latency sample
+    pub fn update_latency(&mut self, address: NetAddress, latency_measurement: u32) -> Result<(), NetAddressError> {
```

Comment:
Suggest u32 -> Duration

Reply:
Shouldn't we take a reference to address here rather than owning it? Same Q for all other methods that do this.

Reply:
Changed to ```Duration```, and changed all the functions to accept the ```NetAddress``` struct as pass by reference